fairlearn/reductions/_moments/conditional_selection_rate.py

```diff
@@ -72,24 +97,29 @@ def load_data(self, X, y, event=None, **kwargs):
 
     def gamma(self, predictor):
         """Calculate the degree to which constraints are currently violated by the predictor."""
-        pred = predictor(self.X)
+        utility_diff = self.utilities[_POSITIVE_UTILITY] - self.utilities[_NEGATIVE_UTILITY]
+        pred = utility_diff * predictor(self.X) + self.utilities[_NEGATIVE_UTILITY]
         self.tags[_PREDICTION] = pred
         expect_event = self.tags.groupby(_EVENT).mean()
         expect_group_event = self.tags.groupby(
             [_EVENT, _GROUP_ID]).mean()
-        expect_group_event[_DIFF] = expect_group_event[_PREDICTION] - expect_event[_PREDICTION]
-        g_unsigned = expect_group_event[_DIFF]
-        g_signed = pd.concat([g_unsigned, -g_unsigned],
+        expect_group_event[_UPPER_BOUND_DIFF] = self.ratio * expect_group_event[_PREDICTION] - \
+            expect_event[_PREDICTION]
+        expect_group_event[_LOWER_BOUND_DIFF] = - expect_group_event[_PREDICTION] \
+            + self.ratio * expect_event[_PREDICTION]
+        g_signed = pd.concat([expect_group_event[_UPPER_BOUND_DIFF],
+                              expect_group_event[_LOWER_BOUND_DIFF]],
                              keys=["+", "-"],
                              names=[_SIGN, _EVENT, _GROUP_ID])
-        self._gamma_descr = str(expect_group_event[[_PREDICTION, _DIFF]])
+        self._gamma_descr = str(expect_group_event[[_PREDICTION, _UPPER_BOUND_DIFF,
+                                                    _LOWER_BOUND_DIFF]])
         return g_signed
 
     # TODO: this can be further improved using the overcompleteness in group membership
     def project_lambda(self, lambda_vec):
         """Return the projected lambda values."""
-        lambda_pos = lambda_vec["+"] - lambda_vec["-"]
-        lambda_neg = -lambda_pos
+        lambda_pos = lambda_vec["+"] - self.ratio * lambda_vec["-"]
+        lambda_neg = - self.ratio * lambda_vec["+"] + lambda_vec["-"]
```

Comment:
This is not correct. I guess I never told you what's expected of `project_lambda`... it needs to return lambda which is guaranteed to lead to the same or higher value of the Lagrangian compared with `lambda_vec` for *all* possible choices of the classifier _h_.

So let's do the following. If `self.ratio != 1.0` then just return `lambda_vec` as given, and when `self.ratio == 1.0` then use the previous code and return `lambda_projected`.

Reply:
In the meantime, I think I figured what the right code should be. Please try this instead of lines 121--122:
```python
lambda_pos = lambda_vec["+"]
lambda_neg = lambda_vec["-"]
lambda_min = np.minimum(lambda_pos, lambda_neg)
lambda_max = np.maximum(lambda_pos, lambda_neg)
lambda_subtract = np.minimum(lambda_max, lambda_min/self.ratio)
lambda_pos = lambda_pos - lambda_subtract
lambda_neg = lambda_neg - lambda_subtract
```
See if things work with the above.

Reply:
I will try this out and get back.
I'll also document the `project_lambda` for ease.

Reply:
This doesn't work right. Example: When I try to run test data with this code for `ErrorRateRatio` with `eps=0.05`, I get `error=0.75` and `disparity=-0.042857` which is incorrect. Ideally, for negative disparity, it should give minimum error.

I have reverted the change to original code for now.

Reply:
This was addressed after today's discussion. Thanks!